### PR TITLE
Setup and migrate to Prisma

### LINKED ISSUE #426: Migrate to Prisma
[Prisma](https://www.prisma.io/) has first-class support for TypeScript and migrations are much nicer than Sequelize.

### PR DESCRIPTION
**Pull request by pondpiu**
Closes #426 

- replace sequelize model with prisma schema
- replace `sequelize.sync()` with prisma migrate
- `migrations/20220101155430_migrate_from_sequelize` to replicate existing database schema (need to be baselined)

What's left:
- modify start script to inject database url and automatically baseline migration if needed

- [x] I updated the changelog

### CODE REVIEW

File: src/utils/db.ts [hunk 1]
```diff
@@ -1,12 +1,3 @@
-import {Sequelize} from 'sequelize-typescript';
-import path from 'path';
-import {DATA_DIR} from '../services/config.js';
-import {FileCache, KeyValueCache, Settings, Shortcut} from '../models/index.js';
+import PrismaModule from '@prisma/client';
```
Comment (codetheweb) [inline src/utils/db.ts:1]:
```suggestion
import {PrismaClient} from '@prisma/client';
```

File: src/utils/db.ts [hunk 2]
```diff
@@ -1,12 +1,3 @@
...
 
-export const sequelize = new Sequelize({
-  dialect: 'sqlite',
-  database: 'muse',
-  storage: path.join(DATA_DIR, 'db.sqlite'),
-  models: [FileCache, KeyValueCache, Settings, Shortcut],
-  logging: false,
-});
+export const prisma = new PrismaModule.PrismaClient();
```
Comment (codetheweb) [inline src/utils/db.ts:3]:
```suggestion
export const prisma = new PrismaClient();
```

File: package.json [hunk 3]
```diff
@@ -21,10 +21,12 @@
     "lint:fix": "eslint 'src/**/*.ts' --fix",
     "clean": "rm -rf dist dts",
     "test": "npm run lint",
-    "build": "tsc",
+    "build": "npm run prisma-generate && tsc",
```
Comment (codetheweb) [inline package.json:24]: The Prisma client should be generated after `yarn install`, so I don't think this is necessary.

**Review by codetheweb — CHANGES_REQUESTED** [review]
Looking good, thank you for working on this!

Main comment so far is that all the models in `schema.prisma` should be named in singular form (i.e. `model FileCaches` -> `model FileCache`).

File: src/utils/db.ts (hunk 1 above)
Comment (pondpiu) [reply src/utils/db.ts:1]: Screen Shot 2565-01-03 at 19 48 18

I got this error when import like that
any suggestions?

**Comment by pondpiu** [comment]
I resolve to renaming db.sqlite to db2.sqlite to distinguish whether `prisma migrate resolve` has been ran yet.
Let me know if you want me to actually read db instead (would prob. need sqlite driver like sqlite3)

Also i stuck on trying to inject `DATABASE_URL` , [dotenv-cli](https://github.com/entropitor/dotenv-cli) doesn't seems to work with variable expansion

File: src/utils/db.ts (hunk 1 above)
Comment (codetheweb) [reply src/utils/db.ts]: Looks like Prisma is only publishing a CommonJS module for now, so we'll leave it as-is.

**Comment by codetheweb** [comment]
I think it's better to migrate the database in-place.

I just changed a bunch of stuff and it's working well for me now, check it out and let me know if it looks good to you. If so, I think it's ready for merge.

File: package.json [hunk 4]
```diff
@@ -24,8 +24,10 @@
     "build": "tsc",
     "build:watch": "tsc --watch",
     "prepack": "npm run clean && npm run build",
-    "start": "node --experimental-json-modules dist/src/index.js",
+    "start": "node --experimental-json-modules dist/src/scripts/prestart.js",
     "dev": "concurrently nodemon 'tsc --watch'",
+    "migrations:generate": "prisma migrate dev",
+    "migrations:run": "prisma migrate deploy",
```
Comment (pondpiu) [inline package.json:30]: These 2 command would fail without `DATABASE_URL` env

File: src/scripts/prestart.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  const client = new Prisma.PrismaClient();
+
+  const spinner = ora('Applying database migrations...').start();
+
+  try {
+    await client.$queryRaw`SELECT COUNT(id) FROM _prisma_migrations`;
+  } catch (error: unknown) {
+    if (error instanceof Prisma.Prisma.PrismaClientKnownRequestError && error.code === 'P2010') {
+      try {
+        await migrateFromSequelizeToPrisma();
```
Comment (pondpiu) [inline src/scripts/prestart.ts:33]: Must not run for new user without db

File: package.json (hunk 4 above)
Comment (codetheweb) [reply package.json]: Good point, fixed in f859f55.

File: src/scripts/prestart.ts (hunk 5 above)
Comment (codetheweb) [reply src/scripts/prestart.ts]: Thanks, fixed in a944927.

**Comment by codetheweb** [comment]
Thanks for the feedback, issues should now be fixed if you want to take another look.

**Comment by pondpiu** [comment]
Looks awesome 👏

**Review by codetheweb — APPROVED** [review]

**Comment by codetheweb** [comment]
Thanks for helping out with this!

**Pull request merged**